Library/Homebrew/completions.rb

```diff
@@ -0,0 +1,79 @@
+# typed: true
+# frozen_string_literal: true
+
+require "utils/link"
+
+# Helper functions for generating shell completions.
+#
+# @api private
+module Completions
+  extend T::Sig
+
+  module_function
+
+  sig { void }
+  def link!
+    write_completions_option "yes"
+    Tap.each do |tap|
+      Utils::Link.link_completions tap.path, "brew completions link"
+    end
+  end
+
+  sig { void }
+  def unlink!
+    write_completions_option "no"
+    Tap.each do |tap|
+      next if tap.official?
+
+      Utils::Link.unlink_completions tap.path
+    end
+  end
+
+  sig { returns(T::Boolean) }
+  def link_completions?
+    read_completions_option == "yes"
+  end
+
+  sig { returns(T::Boolean) }
+  def completions_to_link?
+    shells = %w[bash fish zsh]
+    Tap.each do |tap|
+      next if tap.official?
+
+      shells.each do |shell|
+        return true if (tap.path/"completions/#{shell}").exist?
+      end
+    end
+
+    false
+  end
+
+  sig { params(option: String).returns(String) }
+  def read_completions_option(option: "linkcompletions")
+    HOMEBREW_REPOSITORY.cd do
+      Utils.popen_read("git", "config", "--get", "homebrew.#{option}").chomp
+    end
```

Comment:
I wonder if this logic could/should be generalised more so it can be used more than just here and analytics? Could be "no" or a follow-up PR.

Reply:
Yeah, definitely a good idea. I think it's used often enough that it makes sense. Unless you feel strongly that it should be in a different PR we may as well do it here I think.